plugins/utils/which-key.nix

```diff
@@ -121,153 +475,44 @@ with lib;
-          max = 50;
-        } "min and max width of the columns";
-        spacing = helpers.defaultNullOpts.mkInt 3 "spacing between columns";
-        align = helpers.defaultNullOpts.mkEnumFirstDefault [
-          "left"
-          "center"
-          "right"
-        ] "";
-      };
-
-    ignoreMissing = helpers.defaultNullOpts.mkBool false "enable this to hide mappings for which you didn't specify a label";
-
-    hidden = helpers.defaultNullOpts.mkListOf types.str [
-      "<silent>"
-      "<cmd>"
-      "<Cmd>"
-      "<CR>"
-      "^:"
-      "^ "
-      "^call "
-      "^lua "
-    ] "hide mapping boilerplate";
-
-    showHelp = helpers.defaultNullOpts.mkBool true "show a help message in the command line for using WhichKey";
-
-    showKeys = helpers.defaultNullOpts.mkBool true "show the currently pressed key and its label as a message in the command line";
-
-    triggers = helpers.defaultNullOpts.mkNullable (
-      with types; either (enum [ "auto" ]) (listOf str)
-    ) "auto" "automatically setup triggers, or specify a list manually";
-
-    triggersNoWait =
-      helpers.defaultNullOpts.mkListOf types.str
-        [
-          "`"
-          "'"
-          "g`"
-          "g'"
-          "\""
-          "<c-r>"
-          "z="
-        ]
-        ''
-          list of triggers, where WhichKey should not wait for timeoutlen and show immediately
-        '';
-
-    triggersBlackList =
-      helpers.defaultNullOpts.mkAttrsOf (types.listOf types.str)
-        {
-          i = [
-            "j"
-            "k"
-          ];
-          v = [
-            "j"
-            "k"
-          ];
-        }
-        ''
-          list of mode / prefixes that should never be hooked by WhichKey
-          this is mostly relevant for keymaps that start with a native binding
-        '';
-
-    disable = {
-      buftypes = helpers.defaultNullOpts.mkListOf types.str [ ] "Disabled by default for Telescope";
-      filetypes = helpers.defaultNullOpts.mkListOf types.str [ ] "";
-    };
   };
 
-  config =
+  extraConfig =
     let
-      cfg = config.plugins.which-key;
-      opt = options.plugins.which-key;
-      setupOptions = {
-        plugins = {
-          inherit (cfg.plugins) marks registers spelling;
-          presets = {
-            inherit (cfg.plugins.presets) operators motions;
-            text_objects = cfg.plugins.presets.textObjects;
-            inherit (cfg.plugins.presets)
-              windows
-              nav
-              z
-              g
-              ;
-          };
-        };
-        inherit (cfg) operators;
-        key_labels = cfg.keyLabels;
-        inherit (cfg) motions icons;
-        popup_mappings = {
-          scroll_down = cfg.popupMappings.scrollDown;
-          scroll_up = cfg.popupMappings.scrollUp;
-        };
+      windowIsDefined = any (o: o.isDefined && o.value != null) (attrValues opt.window);
+    in
+    cfg: {
+      plugins.which-key.settings = optionalAttrs windowIsDefined {
```

Comment:
This is what I originally suggested, but it'd be cleaner IMO to use `mkMerge` here:

```
extraConfig = cfg: lib.mkMerge [
  (optionalAttrs windowIsDefined {
    plugins.which-key.settings.window = #omitted
  })
  {/* main config */}
];
```

Otherwise, we end up with an extra `settings` definition, just setting it to `{}`. Not a big deal though, since that'd be merged fine with more useful definitions.

This is definitely a nit, either approach works and it's personal opinion which is best. On a different day I may prefer the current approach, it really is that subjective!